File: Java/Strategy/src/duck/Duck.java
package duck;

import behavior.FlyBehavior;
import behavior.QuackBehavior;

public abstract class Duck {
    protected QuackBehavior quackBehavior;
    protected FlyBehavior flyBehavior;
    protected String TAG;

    public void display() {
        System.out.println(">> " + TAG + " <<");
    }

    public void swim() {
        System.out.println("Swim Swim");
    }

    public void quack() {
        quackBehavior.quack();
    }

    public void fly() {
        flyBehavior.fly();
    }

    public void setQuackBehavior(QuackBehavior quackBehavior){
        this.quackBehavior = quackBehavior;
    }

    public void setFlyBehavior(FlyBehavior flyBehavior) {
        this.flyBehavior = flyBehavior;
    }
}

File: Kotlin/src/main/kotlin/chapter1/Duck.kt
package chapter1

abstract class Duck {
    abstract var flyBehavior: FlyBehavior
    abstract var quackBehavior: QuackBehavior

    abstract fun display()

    fun performFly() {
        flyBehavior.fly()
    }

    fun performQuack() {
        quackBehavior.quack()
    }

    fun swim() {
        println("모든 오리는 물에 뜨죠. 가짜 오리도 물에 떠요.")
    }
}